src/adapters/SwivelAdapter.sol

```diff
@@ -0,0 +1,158 @@
+    /// @notice timestamp from which values flowing through protocol has begun
+    mapping(uint8 => uint256) public periodStart;
+    /// @notice estimated price of ether, set by the admin
+    uint256 public etherPrice = 2_500;
+
+    function approve(address[] calldata a) external {}
+
+    function lend(bytes calldata d) external returns (uint256) {
+        // Parse the calldata into the arguments
+        (
+            uint256[] memory amounts,
+            Swivel.Order[] memory orders,
+            Swivel.Components[] memory components,
+            address pool,
+            uint256 slippage,
+            bool swapFlag
+        ) = abi.decode(
+                d,
+                (
+                    uint256[],
+                    Swivel.Order[],
+                    Swivel.Components[],
+                    address,
+                    uint256,
+                    bool
+                )
+            );
+
+        // Cache a couple oft-referenced variables
+        address underlying = orders[0].underlying;
+        uint256 maturity = orders[0].maturity;
+        address pt = IMarketPlace(marketPlace).markets(underlying, maturity, 1);
+
+        // verify orders are for the same underlying
+        {
+            for (uint256 i = 0; i < orders.length; ) {
+                if (
+                    underlying != orders[i].underlying ||
+                    maturity != orders[i].maturity
+                ) {
+                    revert Exception(0, 0, 0, address(0), address(0)); // TODO: assign exception code
+                }
+
+                unchecked {
+                    i++;
+                }
+            }
+        }
+
+        // get the amount of the orders
+        {
+            uint256 total;
+            for (uint256 i = 0; i < amounts.length; ) {
+                total += amounts[i];
+
+                // extract fee
+                if (i == amounts.length - 1) {
+                    uint256 fee = total / 1000; // todo: fetch feenominator
+                    amounts[i] = amounts[i] - fee;
+                    total = total - fee;
+                }
+
+                unchecked {
+                    i++;
+                }
+            }
+        }
+
+        // store amount of iPTs to be minted to user
+        uint256 received = IERC20(pt).balanceOf(address(this));
+
+        // execute the orders
+        uint256 premium = IERC20(underlying).balanceOf(address(this));
+        ISwivel(protocolAddressess[0]).initiate(orders, amounts, components);
+        premium = IERC20(underlying).balanceOf(address(this)) - premium;
+        received = IERC20(pt).balanceOf(address(this)) - received;
+
+        // Swap the premium for iPTs or return premium to the sender
+        if (swapFlag) {
+            received += swap(pool, premium, slippage);
+        } else {
+            premiums[underlying][maturity] += premium;
+            received += premium;
+        }
+
+        return received;
```

Comment:
So after reviewing the whole PR, the proxy returns the raw received amount, after having been passed the raw underlying input amount -- So I dont think a fee is being taken?

Reply:
Yup, the fee stuff was definitely confusing (calculating it partially in the adapter and Lender). I have changed it so that it is now fully handled by the adapter.